Authentication: clear text, md5 and without password

orlangure opened the pull request:
[Task](https://app.asana.com/0/573768021211412/768621478822192/f)

This PR adds `authenticator` interface and its 3 implementations:
- without password (to keep things as they are now)
- clear text password
- md5 hashed password

Last two implementations use dependency injection and delegate tasks like "get password for user" to provided objects. This also helps in testing (`auth.go` is almost completely covered with unit tests).

This PR also modifies `err` type: it adds a new field for severity (auth errors are `FATAL`; `ERROR` is not enough for them)

Currently, `sess.go` uses `authenticationNoPassword`. An example of md5 looks like the following:

```go
// password provider - implement when available
pp := &md5ConstantPasswordProvider{password: []byte("password")}

// authenticator that accepts current session as rw (readWriter),
// session args and password provider
a := &authenticationMD5{rw: s, args: s.Args, pp: pp}

authResponse, err := a.authenticate()
if err != nil {
	return s.Write(errMsg(WithSeverity(err, "FATAL")))
}
```

Commit:
Add severity to custom errors

Commit:
Use severity in error messages

Commit:
Mark session as initialized before auth

Commit:
Implement authentication

This commit adds authenticator interface with 3 implementations:
- no password (instant authentication)
- clear text password
- md5 hashed password

One of these authenticators can be selected in sess.go

Commit:
Use predefined authOK message

Commit:
Move authOKMsg() to auth.go

Commit:
Add test for authOKMsg

Commit:
Pull all dependencies in travis

Commit:
trigger travis

aviaoh reviewed (changes requested):
as agreed, waiting for changes in this pr to allow authenticator to store hashed md5 passwords.

Commit:
Expect md5 hash of user/pass from passwordProvider

Commit:
Merge branch 'master' of https://github.com/panoplyio/pgsrv into authentication

aviaoh commented on errors.go:
```diff
@@ -26,18 +26,31 @@ import (
+		return nil
+	}
+
+	e := fromErr(err)
```
now that we have this field, please add it to `fromErr`

aviaoh commented on msg_query.go:
```diff
@@ -106,6 +106,14 @@ func errMsg(err error) msg {
 		"M": err.Error(),
 	}
 
+	// severity
```
now that we have new field, please move default "ERROR" severity from this file (line 104) to initialization in fromErr function

orlangure replied:
@aviaoh, `errMsg` accepts any error, and if we move the default to `fromErr`, we risk having errors without severity. The same applies for error code: the default is here, although `Code()` is used in `fromErr` as well.

aviaoh commented on sess.go:
```diff
@@ -92,8 +92,16 @@ func (s *session) Serve() error {
+	err = s.Write(authResponse)
 	if err != nil {
 		return err
 	}
```
👍 
nice and clean :)

aviaoh commented on sess.go:
```diff
@@ -92,8 +92,16 @@ func (s *session) Serve() error {
+
 	// handle authentication.
-	err = s.Write(authOKMsg())
+	a := &authenticationNoPassword{}
```
1. please add following comment (we will get to it after handling md5 store on epsilon):
```
// TODO: replace with actual pre-configured authenticator
```

2. the authenticator should be defined on Server instance, not on each and every session creation. see comment on authenticator interface

aviaoh commented on sess.go:
```diff
@@ -92,8 +92,16 @@ func (s *session) Serve() error {
+	a := &authenticationNoPassword{}
+	authResponse, err := a.authenticate()
+	if err != nil {
+		return s.Write(errMsg(WithSeverity(err, "FATAL")))
```
wondering if the authenticator should `return WithSeverity(err, "FATAL")` or if it should be done here.

are all authenticator consider this as fatal? what if authenticator wants to return different severity? (like ERROR/PANIC)?

we don't have to address it now, just wondering what did you see in postgres code :)

aviaoh replied:
after revisiting authenticate code, i think it should return only error without ok/err message, and we should simply terminate the session in case of error.

the way i see it, only authenticator know if err msg should be sent.
for example, if authenticate gets error in rw.Read - no reason to write fatal error back to client.
instead, allow authenticate to send ok message and here only `return err` as any other case here...

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+// It requires a messageReadWriter implementation to communicate with the client,
+// passwordProvider implementation to verify that the provided password is correct,
+// and a map of arguments that were sent at the beginning of the session (user, database, etc)
+type authenticationMD5 struct {
```
please rename to `md5Authenticator`

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+}
+
+// authenticationMD5 is an authenticator that requests and accepts an MD5 hashed password
+// from the client.
```
please rephrase as `authenticationMD5 requests and accepts an MD5 hashed password from the client.`

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+// authenticator interface defines objects able to perform user authentication
+// that happens at the very beginning of every session.
+type authenticator interface {
+	authenticate() (msg, error)
```
this signature will require us to create authenticator for each and every authenticate request.

what do you think about changing it to the following?
```
authenticate(rw messageReadWriter, args map[string]interface{}) error
```

aviaoh replied:
see below.. what do you think about returning only error here without a message?

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+	passwordRequest := msg{
+		'R',
+		0, 0, 0, 12,
+		0, 0, 0, 5,
```
please add comments of `// length` and `// md5 auth type`

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+
+	if m.Type() != 'p' {
+		return msg{},
+			fmt.Errorf("expected password response, got message type %c", m.Type())
```
please use const string (here and in md5)

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+	pp   passwordProvider
+}
+
+func (a *authenticationClearText) authenticate() (msg, error) {
```
a lot of code duplication here and in authenticationMD5. please try to reuse the code

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+// It requires a messageReadWriter implementation to communicate with the client,
+// passwordProvider implementation to verify that the provided password is correct,
+// and a map of arguments that were sent at the beginning of the session (user, database, etc)
+type authenticationClearText struct {
```
please rename to clearTextAuthenticator

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,184 @@
+}
+
+// authenticationClearText is an authenticator that requests and accepts a clear text password
+// from the client. It is not recommended to use it for security reasons.
```
please rephrase as `authenticationClearText requests and accepts a clear text password...`

aviaoh reviewed (changes requested):
Cool changes :)

please see comments below regarding architecture..

Commit:
Use Severity() in fromErr()

Commit:
Add TODO comment for authentication

Commit:
Rename authenticators

Commit:
Add comments to byte messages

Commit:
Modify authenticator interface to return bool+err

All the communication with clients happens in authenticate() method.
It returns only if everything is ok (auth succeeded) and if there was an
error somewhere in the process. If the auth is not ok, or there was an
error, the session must be terminated by the called.

orlangure commented:
@aviaoh, addressed almost all your comments.
I couldn't think of a way to reduce code duplication in a sane way. Only errors are the same, and it doesn't feel right to extract these errors as they should be sent as messages from `authenticate` itself anyway.

Commit:
Use const for repeated strings

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,192 @@
+)
+
+const expectedPasswordMessage = "expected password response, got message type %c"
+const passwordDidNotMatch = "Password does not match for user \"%s\""
```
please rename to `errExpectedPassword` and `errWrongPassword`

aviaoh commented on sess.go:
```diff
@@ -92,12 +92,20 @@ func (s *session) Serve() error {
 	// handle authentication.
-	err = s.Write(authOKMsg())
+	// TODO: replace with an actual pre-configured authenticator
+	a := &noPasswordAuthenticator{}
```
the authenticator should be defined on Server instance, not on each and every session creation (from previous review cycle)

orlangure replied:
@aviaoh I thought to move it when we have a real authenticator. Will move now.

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,192 @@
+	// Authentication errors as well as welcome messages are sent by this function,
+	// so there is no need for the caller to send these. It is caller's responsibility
+	// though to terminate the session in case that the first return value is false
+	// OR if the second returned value is not nil.
```
👍

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,192 @@
+	// so there is no need for the caller to send these. It is caller's responsibility
+	// though to terminate the session in case that the first return value is false
+	// OR if the second returned value is not nil.
+	authenticate(rw msgReadWriter, args map[string]interface{}) (bool, error)
```
@orlangure - why do we need to return two values here?
why not return const fmt.Errorf("authentication failed") here instead of false? is there any value of two cases that indicate an error?

orlangure replied:
@aviaoh my point is that a failed auth is not the same as an error that occured somewhere in the process. Think of it as that if the second value exists (`err != nil`), then there is no definitive decision whether or not the user can access the system. If the second value is `nil` (no error), we know exactly whether or not to let the user in.

aviaoh replied:
but we are not letting the user in on both cases... 
authentication can failed with any error, either connection error or wrong password.

what is the difference from caller point of view?

aviaoh commented on auth_test.go:
```diff
@@ -0,0 +1,250 @@
+	})
+}
+
+// mockMessageReadWriter implements messageReadWriter and outputs the provided output
```
it's [more stub than a mock](https://martinfowler.com/articles/mocksArentStubs.html), but cool tests 🤓

aviaoh commented on errors.go:
```diff
@@ -4,6 +4,9 @@ import (
 )
 
+// FATAL error terminates session
+const FATAL = "FATAL"
```
should it be public?
please rename to `fatalSeverity`

aviaoh commented on sess.go:
```diff
@@ -92,12 +92,20 @@ func (s *session) Serve() error {
 
+	if !ok {
+		return fmt.Errorf("authentication failed")
+	}
```
redundant? see comments regarding two returned values

aviaoh reviewed (changes requested):
cool cool cool :)

requesting changes for finalize the signature and create single authenticator per server

Commit:
Rename error constants

Commit:
Make severity const private

Commit:
Move autheticator into Server instance

Commit:
Simplify authenticator signature

aviaoh commented on auth_test.go:
```diff
@@ -0,0 +1,243 @@
+
+		require.Equal(t, passwordRequest, rw.messages[0])
+		require.True(t, bytes.Contains(rw.messages[1], fatalMarker))
+		require.NoError(t, err)
```
??
🤔 

(same for all invalid tests)

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,189 @@
+	if !bytes.Equal(expectedPassword, actualPassword) {
+		err = fmt.Errorf(errWrongPassword, user)
+		m := errMsg(WithSeverity(fromErr(err), fatalSeverity))
+		return rw.Write(m)
```
you should return error here.. not write result

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,189 @@
+	if m.Type() != 'p' {
+		err = fmt.Errorf(errExpectedPassword, m.Type())
+		m := errMsg(WithSeverity(fromErr(err), fatalSeverity))
+		return rw.Write(m)
```
you should return error here.. not write result

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,189 @@
+	if m.Type() != 'p' {
+		err = fmt.Errorf(errExpectedPassword, m.Type())
+		m := errMsg(WithSeverity(fromErr(err), fatalSeverity))
+		return rw.Write(m)
```
you should return error here.. not write result

aviaoh commented on auth.go:
```diff
@@ -0,0 +1,189 @@
+	if !bytes.Equal(expectedHash, actualHash) {
+		err = fmt.Errorf(errWrongPassword, user)
+		m := errMsg(WithSeverity(fromErr(err), fatalSeverity))
+		return rw.Write(m)
```
you should return error here.. not write result

aviaoh commented on srv.go:
```diff
@@ -9,15 +9,16 @@ import (
 // executing SQL commands (see Execer).
 func New(queryer Queryer) Server {
-	return &server{queryer}
+	return &server{queryer, &noPasswordAuthenticator{}}
```
please add `// TODO: replace with an actual pre-configured authenticator` :)

aviaoh reviewed (changes requested):
looks good :)

see bugs and tests that should be fixed

Commit:
Return auth errors and check them in tests

Commit:
Add TODO to replace noPassword with md5

aviaoh reviewed (approved):
Looks great :)
